x/blobstream/types/validator.go

```diff
@@ -112,11 +113,12 @@ func EVMAddrLessThan(e common.Address, o common.Address) bool {
 // increases by 1% due to inflation, we shouldn't have to generate a new
 // validator set, after all the validators retained their relative percentages
 // during inflation and normalized Blobstream power shows no difference.
-func (ibv InternalBridgeValidators) PowerDiff(c InternalBridgeValidators) float64 {
+func (ibv InternalBridgeValidators) PowerDiff(
+	c InternalBridgeValidators) sdk.Dec {
 	powers := map[string]int64{}
 	// loop over ibv and initialize the map with their powers
 	for _, bv := range ibv {
-		powers[bv.EVMAddress.Hex()] = int64(bv.Power)
+		powers[bv.EVMAddress.Hex()] = int64(bv.Power) // TODO: overflow?
```

Comment:
The validator powers are of type `uint64`, and the current conversion could potentially lead to overflow. Upon reviewing the [`validateBasics`](https://github.com/celestiaorg/celestia-app/blob/6683a049b95729c2d055c8e08e3c81d9b7bb7914/x/blobstream/types/validator.go#L57) function, it appears that there is no range check performed on the power value. It would be prudent to confirm whether there might be any validators with powers falling within the range of [`math.MaxInt64`, `math.MaxUint64`]. Such power values could indeed cause an overflow issue in the code line mentioned above. (For example the power value of `math.MaxUint64` becomes `-1` when converted to `int64`)

Reply:
good catch :+1: But I think if the state machine got to a state where this overflows, then something is seriously messed up with the staking module. In fact, values like those will not be accepted by the staking module, then we normalize those values before this step. I think that we can use a validation at this level as a last line of defense, but reaching it is super unlikely if I understand corerctly

Reply:
> values like those will not be accepted by the staking module

So, does it mean that there will never be a case in which a validator's power reside in this range [math.MaxInt64, math.MaxUint64]? and it is guaranteed by the staking module?

Reply:
yes, the validator power is using `int64`, so no overflow would happen. However, I found something interesting:
https://github.com/celestiaorg/cosmos-sdk/blob/main/x/staking/types/validator.go#L369-L372

```
// PotentialConsensusPower returns the potential consensus-engine power.
func (v Validator) PotentialConsensusPower(r math.Int) int64 {
	return sdk.TokensToConsensusPower(v.Tokens, r)
}
```

The power is being reduced by a factor of:
```
// DefaultPowerReduction is the default amount of staking tokens required for 1 unit of consensus-engine power
var DefaultPowerReduction = NewIntFromUint64(1000000)
```

However, the tokens amount is represented by an int256... So, there might be overflow there when going from tokens to consensus-engine power unless there is the assumption that the staked tokens wouldn't be that much.

cc @evan-forbes any idea?

Reply:
sorry mind pointing out where's the int256 to save me a sec :sweat_smile: 

this is good to keep in the back of our heads, especially for networks such as arabica where the supply and voting power is getting rather insane. for normal networks tho the assumption of staying within the max and min values appears safe afaiu (please correct me if I'm wrong I didn't deep dive the code yet)

Reply:
In the [`validator`](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/x/staking/types/staking.pb.go#L314-L339) struct, the delegated tokens are represented using [`Tokens`](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/x/staking/types/staking.pb.go#L324C26-L324C26) which are [`Int`](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/types/math.go#L14)s which are [`sdkmath.Int`](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/math/int.go#L72-L77). 

Then, these are used when calculating the validator delegation and power in [here](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/x/staking/keeper/val_state_change.go#L169) after getting the `consensus-engine power` which would [panic](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/math/int.go#L156-L161) when converting from `sdkmath.Int` to `int64` if the amount is out of bounds.